hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ServiceInfoProvider.java

```diff
@@ -0,0 +1,145 @@
+import org.apache.hadoop.hdds.security.exception.SCMSecurityException;
+import org.apache.hadoop.hdds.security.x509.certificate.client.CertificateClient;
+import org.apache.hadoop.hdds.security.x509.certificate.utils.CertificateCodec;
+import org.apache.hadoop.ozone.om.helpers.ServiceInfoEx;
+import org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol;
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.security.cert.X509Certificate;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyList;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * A helper class to handle the construction of
+ * {@link org.apache.hadoop.ozone.om.helpers.ServiceInfoEx} objects,
+ * and refresh of the things that are necessary for constructing it.
+ */
+public final class ServiceInfoProvider {
+
+  private static final Logger LOG = getLogger(ServiceInfoProvider.class);
+
+  private final OzoneManagerProtocol om;
+  private final CertificateClient certClient;
+
+  private String caCertPem;
+  private List<String> caCertPemList;
+
+  public ServiceInfoProvider(
+      OzoneConfiguration config,
+      OzoneManagerProtocol om,
+      CertificateClient certClient
+  ) throws IOException {
+    this(config, om, certClient, false);
+  }
+
+  public ServiceInfoProvider(
+      OzoneConfiguration config,
+      OzoneManagerProtocol om,
+      CertificateClient certClient,
+      boolean skipInitializationForTesting
+  ) throws IOException {
+    this.om = om;
+    if (new SecurityConfig(config).isSecurityEnabled()
+        && !skipInitializationForTesting) {
+      this.certClient = certClient;
+      Set<X509Certificate> certs = getCACertificates();
+      caCertPem = toPEMEncodedString(newestOf(certs));
+      caCertPemList = toPEMEncodedStrings(certs);
+      this.certClient.registerRootCARotationListener(onRootCAChange());
+    } else {
+      this.certClient = null;
+      caCertPem = null;
+      caCertPemList = emptyList();
+    }
+  }
+
+  private Function<List<X509Certificate>, CompletableFuture<Void>>
+      onRootCAChange() {
+    return certs -> {
+      CompletableFuture<Void> returnedFuture = new CompletableFuture<>();
+      try {
+        caCertPem = toPEMEncodedString(newestOf(certs));
+        caCertPemList = toPEMEncodedStrings(certs);
+        returnedFuture.complete(null);
+      } catch (Exception e) {
+        LOG.info("Unable to refresh cached PEM formatted CA " +
+            "certificates.", e);
+        returnedFuture.completeExceptionally(e);
+      }
+      return returnedFuture;
+    };
+  }
+
+  public ServiceInfoEx provide() throws IOException {
+    return new ServiceInfoEx(om.getServiceList(), caCertPem, caCertPemList);
```

Comment:
Need a way to synchronize the fetch and update of caCertPem and caCertPermList.

Reply:
The easiest seems to synchronize on this when the values are updated in onRootCAChange, and then assign these to variables in a synchronized block so that we can provide a copy to the ServiceInfoEx constructor, and we do not need to care about further synchronization if I think it correclty. Adjusted the code.